### ISSUE TITLE
QrPainter.toImageData generate a bad format QR code

### ISSUE THREAD

**Issue by oom-** [theyakka/qr.flutter#117]
**Describe the bug**
QrPainter's function called `toImageData` return QrCode with cropped pixels on border 

**To Reproduce**
```dart
 String testtext =
          "Contrary to popular belief, Lorem Ipsum is not simply random text. It has roots in a piece of classical Latin literature from 45 BC, making it over 2000 years old. Richard McClintock, a Latin professor at Hampden-Sydney College in Virginia, looked up one of the more obscure Latin words, consectetur, from a Lorem Ipsum passage, and going through the cites of the word in classical literature, discovered the undoubtable source. Lorem Ipsum comes from sections 1.10.32 and 1.10.33 of 'de Finibus Bonorum et Malorum' (The Extremes of Good and Evil) by Cicero, written in 45 BC. This book is a treatise on the theory of ethics, very popular during the Renaissance. The first line of Lorem Ipsum, 'Lorem ipsum dolor sit amet..', comes from a line in section 1.10.32.\nThe standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested. Sections 1.10.32 and 1.10.33 from 'de Finibus Bonorum et Malorum' by Cicero are also reproduced in their exact original form, accompanied by English versions from the 1914 translation by H. Rackham.";
      ByteData qrbytes = await QrPainter(data: testtext, version: QrVersions.auto).toImageData(1000);
      writeToFile(qrbytes, Profile.directory.path + 'test.png');
```

**Expected behavior**
The QRCode generated to be valid without cut on the edges:
![image](https://user-images.githubusercontent.com/11071671/100009671-7dfaca00-2dcf-11eb-9f83-64f5ecc428d2.png)

**Screenshots**
![image](https://user-images.githubusercontent.com/11071671/100009258-eb5a2b00-2dce-11eb-806d-c3134ee86f4d.png)

As you can see the pixels are not full

**Smartphone (please complete the following information):**
 - Device: All
 - OS: Android

**Comment by braysonjohn148** [comment]
Any fix on this. I cannot scan QR Code made by this package. But if i try random QR codes from google. They work

**Comment by oom-** [comment]
I fixed it like this: 
* "Gapless: true"
* Bigger size as parameter

```js
ByteData qrbytes = await QrPainter(data: getQrContent(), gapless: true, version: QrVersions.auto).toImageData(1200);
```

**Comment by iandis** [comment]
As long as I observe, this bug occurs because:
- `_PaintMetrics._innerContentSize` (which is used by `QrPainter.paint`) often produces different size than our requested size.
- Flutter's `Canvas` allows painting outside of the defined size.
- Flutter's `Picture.toImage` automatically discards pixels outside of the defined size.

The fix that works for me:
```dart
double _getActualQRCodeSize({
    required String text,
    required double size,
    required bool gapless,
    required int errorCorrectLevel,
  }) {
    // These calculations are from `_PaintMetrics._calculateMetrics`
    final double gap = gapless ? 0 : 0.25;
    final QrCode qrCode = QrCode.fromData(
      data: text,
      errorCorrectLevel: errorCorrectLevel,
    );

    final double gapTotal = (qrCode.moduleCount - 1) * gap;
    
    final double actualQRCodeSize;
    double pixelSize = (size - gapTotal) / qrCode.moduleCount;
    pixelSize = (pixelSize * 2).roundToDouble() / 2;

    actualQRCodeSize = (pixelSize * qrCode.moduleCount) + gapTotal;

    return actualQRCodeSize;
}

const bool gapless = true;
const int errorCorrectLevel = QrErrorCorrectLevel.M;
const String data = 'abcdefgh12uhsdhfsdjfhiweuhfewhifuw';
const double size = 400;
final double actualQRCodeSize = _getActualQRCodeSize(
      text: data,
      size: size,
      gapless: gapless,
      errorCorrectLevel: errorCorrectLevel,
);

final ByteData? imageByte = await QrPainter(
      data: data,
      version: QrVersions.auto,
      gapless: gapless,
      color: const Color(0xFF000000),
      emptyColor: const Color(0xFFFFFFFF),
      errorCorrectionLevel: errorCorrectLevel,
).toImageData(actualQRCodeSize); // <- we don't use `size` anymore to convert qr to imagedata
```

**Comment by lukef** [comment]
Fixed in 5.0.
